bpo-46752: Introduce task groups in asyncio

gvanrossum opened the pull request:
This is EdgeDB's TaskGroup class, adapted for Python 3.11.

In the individual commits you can see how I evolved this from the version in EdgeDB.

Here's a to-do list:

- [x] Figure out why in `test_taskgroup_14` I get a nested `ExceptionGroup`, while the original got a flat MultiError.
- [x] Fix the test framework's complaint about the event loop policy being changed.
- [x] **Design question:** Do we need `TaskGroupError`?  Implementing it correctly is awkward. We could just raise `BaseExceptionGroup`.
  - ~If we decide to keep it, we need to fix it to do the hack where it inherits from `BaseException` iff at least one of the exceptions does. We also need to override `derive()`.~
- [x] Add tests for handling `BaseException`, ~`KeyboardInterrupt` and `SystemExit`~.
- [x] ~Make `test_taskgroup_21` work? (It wasn't working in EdgeDB either, so it was crudely disabled.)~
- [x] Add a public API to `tasks.py` to replace `__cancel_requested__`, and get rid of the monkey-patching.
- [x] **Design question:** Should we allow creating new tasks while `__aexit__()` is already waiting?
- [ ] Add a test showing the need for the `.uncancel()` call in `__aexit__()` (currently on line 97). Dropping that line does not cause any tests to fail.
- [ ] Ensure the taskgroup tests are run with the C and Python Task implementations.
- [ ] Rename tests to have meaningful names.
- [ ] I have a few ideas for minor cleanups that I will do later.
- [x] What should we do with the copyright header? @1st1.
- [x] Create a bpo issue.
- [x] Add a NEWS blurb.
- [ ] Documentation and What's New entry (in a separate PR, probably).

I also learned a few things about the ergonomics of `ExceptionGroup`:

- The `BaseException` hacks are clever but awkward to replicate in a subclass.
- The `str()` of an EG is rather sparse (Yury's version adds the number of sub-exceptions and their types).
- Yury's version also has a convenience method that returns the set of sub-exception types.
- Unsure about the nesting issue (see above).

CC: @iritkatriel, @1st1

Co-authored-by: Yury Selivanov <yury@edgedb.com>
Co-authored-by: Andrew Svetlov <andrew.svetlov@gmail.com>

<!-- issue-number: [bpo-46752](https://bugs.python.org/issue46752) -->
https://bugs.python.org/issue46752
<!-- /issue-number -->

gvanrossum committed:
Integrate task groups from EdgeDb

My plan is roughly the following:

- [x] Copy the files from EdgeDb without modifications
      (named following asyncio conventions)
- [ ] Bang on the tests until they will run
- [ ] Bang on the tests until they pass
- [ ] Remove pre-3.11 compatibility code
- [ ] Switch from MultiError to ExceptionGroup
- [ ] Other cleanup
- [ ] Add a public API to tasks.py to replace `__cancel_requested__`

gvanrossum committed:
Make test_taskgroups.py run and pass

Two remaining issues:

- [ ] _test_taskgroup_21 doesn't work (it didn't in EdgeDb either)
- [ ] the test framework complains about a change to the event loop policy

gvanrossum committed:
Rename taskgroup to taskgroups in the test code

gvanrossum committed:
Export TaskGroup from asyncio; remove __future__ import

gvanrossum committed:
Only keep the newest  _is_base_error() and _task_cancel()

gvanrossum committed:
Get rid of MultiError in favor of ExceptionGroup

This required some changes to the tests since EdgeDb's MultiError
has some ergonomic conveniences that ExceptionGroup doesn't:

- A helper method to get the types of the exceptions
- It puts the number and types of the exceptions in the message

Also, in one case (test_taskgroup_14) an EG nested inside another
EG was raised, whereas the original code just raised one EG.
This remains to be investigated.

gvanrossum committed:
Add TaskGroupError to __all__

gvanrossum committed:
Avoid DeprecationWarning: There is no current event loop

gvanrossum committed:
Prevent warning "test altered the execution environment"

gvanrossum committed:
Get rid of custom TaskGroupError

Just use [Base]ExceptionGroup

gvanrossum committed:
Update comments explaining why test 21 doesn't work

gvanrossum committed:
Add tests showing that 'plain' BaseExceptions work

gvanrossum committed:
Allow creating new tasks while __aexit__ is waiting

We stop when there are no unfinished tasks,
and then no new tasks can be created.

(TODO: more thorough testing of edge cases?)

gvanrossum committed:
Add an API to Task to manage 'cancel_requested' flag

This means we no longer have to monkey-patch the parent task.

It does introduce new semantics for task cancellation:
When a task is cancelled, further attempts to cancel it
have *no* effect unless the task calls self.uncancel().

Borrowed from GH-31313 by @asvetlov.

gvanrossum committed:
Add tests for .cancelling() and .uncancel()

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into taskgroups

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

1st1 commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,249 @@
+            errors = self._errors
+            self._errors = None
+
+            me = BaseExceptionGroup('unhandled errors in a TaskGroup', errors)
```
Should this be `BaseExceptionGroup` only if `any(not isinstance(e, Exception) for e in errors)`, and `ExceptionGroup` otherwise?

gvanrossum replied:
Not needed, the BaseExceptionGroup design automatically handles that (it's in the PEP :-).

gvanrossum committed:
Replace EdgeDb copyright with a simpler attribution

Yury, the author, re-licenses the code under PSFL,
and since he's signed Python's contributor agreement he has no problem
assigning his copyright on the implementation and tests to the PSF.

asvetlov committed:
Use task.cancelling() in task repr instead of access to private attributes

asvetlov commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,233 @@
+        self._entered = True
+
+        if self._loop is None:
+            self._loop = asyncio.get_running_loop()
```
```suggestion
            self._loop = events.get_running_loop()
```

asvetlov commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,233 @@
+        if self._loop is None:
+            self._loop = asyncio.get_running_loop()
+
+        self._parent_task = asyncio.current_task(self._loop)
```
```suggestion
        self._parent_task = tasks.current_task(self._loop)
```

asvetlov commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,233 @@
+                self._base_error is None):
+            self._base_error = exc
+
+        if et is asyncio.CancelledError:
```
```suggestion
        if et is exceptions.CancelledError:
```

asvetlov commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,233 @@
+            #        g.create_task(...)
+            #        await ...  # <- CancelledError
+            #
+            if et is asyncio.CancelledError:
```
```suggestion
            if et is exceptions.CancelledError:
```

asvetlov commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,233 @@
+
+            try:
+                await self._on_completed_fut
+            except asyncio.CancelledError as ex:
```
```suggestion
            except exceptions.CancelledError as ex:
```

asvetlov commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,233 @@
+            # request now.
+            raise propagate_cancellation_error
+
+        if et is not None and et is not asyncio.CancelledError:
```
```suggestion
        if et is not None and et is not exceptions.CancelledError:
```

asvetlov commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,233 @@
+
+__all__ = ["TaskGroup"]
+
+import asyncio
```
```suggestion
```

asvetlov commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,233 @@
+import traceback
+import types
+import weakref
+
```
```suggestion

from . import events
from . import exceptions
from . import tasks

```

asvetlov reviewed (commented):
Changed `import asyncio` to `from . import <submodule>` to follow asyncio convention for own modules import.

asvetlov reviewed (approved):
LGTM

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+class TestTaskGroup(unittest.IsolatedAsyncioTestCase):
+
+    async def asyncSetUp(self):
+        self.loop = asyncio.get_event_loop()
```
```suggestion

```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+            NUM += 10
+
+        with self.assertRaises(ExceptionGroup) as cm:
+            await self.loop.create_task(runner())
```
```suggestion
            await asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+        # The 3 foo1 sub tasks can be racy when the host is busy - if the
+        # cancellation happens in the middle, we'll see partial sub errors here
+        with self.assertRaises(ExceptionGroup) as cm:
+            await self.loop.create_task(runner())
```
```suggestion
            await asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                for _ in range(5):
+                    g.create_task(foo())
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                    NUM += 10
+                    raise
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                except asyncio.CancelledError:
+                    raise
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                    except asyncio.CancelledError:
+                        raise
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                    except asyncio.CancelledError:
+                        raise
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                async with taskgroups.TaskGroup(name='g2') as g2:
+                    g2.create_task(crash_after(0.2))
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                async with taskgroups.TaskGroup(name='g2') as g2:
+                    g2.create_task(crash_after(0.1))
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                    await asyncio.sleep(0.5)
+                    raise
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                    raise
+
+        async def runner():
+            t = self.loop.create_task(nested_runner())
```
```suggestion
            t = asyncio.create_task(nested_runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+            t = self.loop.create_task(nested_runner())
+            await t
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                    NUM += 10
+                    raise
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                    # this weird case.
+                    raise MyExc
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                g.create_task(crash_soon())
+                await nested()
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                g.create_task(foo1())
+                g.create_task(foo2())
+
+        r = self.loop.create_task(runner())
```
```suggestion
        r = asyncio.create_task(runner())
```

asvetlov reviewed (commented):
`self.loop` is not required if `asyncio.create_task()` shortcut is used.

asvetlov commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,698 @@
+                g1.create_task(crash_after(0.2))
+
+                async with taskgroups.TaskGroup(name='g2') as g2:
+                    g2.create_task(crash_after(0.1))
```
This code seems not robust; it can fail on a slow machine: https://github.com/python/cpython/runs/5201217606?check_suite_focus=true.
Perhaps waiting for `asyncio.Event` (or future) is better than constant small sleeps (but sleeps are better readable, sure).

gvanrossum replied:
Let me fix that by making the longer sleep longer -- that sleep will be interrupted so it doesn't matter how long it is. Let me know if you think that's not the right way to fix this test.

gvanrossum commented on Lib/test/test_asyncio/test_tasks.py:
```diff
@@ -506,8 +506,10 @@ async def task():
             self.assertTrue(t.cancel())
             self.assertTrue(t.cancelling())
+            self.assertIn(" cancelling ", repr(t))
             self.assertFalse(t.cancel())
```
I'd like to also test that the cancelled bit stays on even if the task catches CancelledError and sleeps some more.

gvanrossum replied:
(I tried to implement this but couldn't get it to work. The task always seems to run through to completion?)

asvetlov replied:
I'll try to write such a test later. May take a few days, depending on my other duties.

gvanrossum committed:
Change the internal imports

gvanrossum committed:
Avoid needing self.loop in test

gvanrossum committed:
Make test 14 more robust

(By making the sleep in the outer task longer.)

1st1 commented on Lib/asyncio/taskgroups.py:
```diff
@@ -0,0 +1,235 @@
+# Adapted with permission from the EdgeDb project.
```
```suggestion
# Adapted with permission from the EdgeDB project.
```

1st1 commented on Lib/test/test_asyncio/test_taskgroups.py:
```diff
@@ -0,0 +1,694 @@
+# Adapted with permission from the EdgeDb project.
```
```suggestion
# Adapted with permission from the EdgeDB project.
```

gvanrossum replied:
Dang I saw this the second I merged. Will fix.

Commit:
Update Lib/asyncio/taskgroups.py

Commit:
Update Lib/test/test_asyncio/test_taskgroups.py

1st1 reviewed (approved):
@gvanrossum I think this is ready. With this merged we'll need to update the docs in a few places to de-prioritize `asyncio.gather()` and steer people towards TaskGroups.
